File: mcp_server_development/python/src/session12/auth_middleware.py
"""ASGI 認証ミドルウェア（Python 版）

TypeScript 版は http.RequestListener を包みましたが、Python SDK は HTTP サーバーを
内部で組み立てるため、包む場所が ASGI アプリになります。やっていることは同じです。

  ① Protected Resource Metadata を無認証で公開する
  ② Authorization ヘッダーを検証して 401
  ③ ベーススコープを見て 403
  ④ 認証文脈を scope["state"] に載せて下流へ渡す
"""

from __future__ import annotations

import json
import sys
from collections.abc import Awaitable, Callable
from typing import Any

from tokens import SCOPE_DOCS_READ, SUPPORTED_SCOPES, AuthError, verify_token

Scope = dict[str, Any]
Receive = Callable[[], Awaitable[dict[str, Any]]]
Send = Callable[[dict[str, Any]], Awaitable[None]]


class BearerAuthMiddleware:
    def __init__(
        self,
        app: Callable[[Scope, Receive, Send], Awaitable[None]],
        *,
        secret: str,
        issuer: str,
        audience: str,
        endpoint: str = "/mcp",
        base_scope: str = SCOPE_DOCS_READ,
        skip_audience: bool = False,
    ) -> None:
        self.app = app
        self.secret = secret
        self.issuer = issuer
        self.audience = audience
        self.base_scope = base_scope
        self.skip_audience = skip_audience
        base = "/.well-known/oauth-protected-resource"
        # RFC 9728 のパス挿入と、パス無しの位置の両方で応答する
        self.metadata_paths = (base, f"{base}{endpoint}")
        origin = audience.split(endpoint)[0]
        self.metadata_url = f"{origin}{base}{endpoint}"
        self.metadata_body = json.dumps(
            {
                "resource": audience,
                "authorization_servers": [issuer],
                "scopes_supported": list(SUPPORTED_SCOPES),
                "bearer_methods_supported": ["header"],
                "resource_name": "社内ドキュメント検索 MCP サーバー（Python 版）",
            },
            ensure_ascii=False,
        ).encode("utf-8")

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        if scope.get("type") != "http":
            await self.app(scope, receive, send)
            return

        path = str(scope.get("path", ""))
        if path in self.metadata_paths:
            await self._send(send, 200, self.metadata_body, [])
            return

        token = self._bearer_token(scope)
        if token is None:
            await self._challenge(send, 401, None, "An access token is required")
            return
        try:
            context = verify_token(
                token,
                secret=self.secret,
                issuer=self.issuer,
                audience=self.audience,
                skip_audience=self.skip_audience,
            )
        except AuthError as error:
            # 理由はログに残し、クライアントには一般化した文面だけ返す
            print(f"[mcp-auth] 認証失敗（{error.reason}）: {path}", file=sys.stderr)
            await self._challenge(send, 401, "invalid_token", "The access token is invalid")
            return

        if self.base_scope not in context.scopes:
            print(
                f"[mcp-auth] 認可失敗（insufficient_scope）: sub={context.subject}"
                f" token={context.fingerprint}",
                file=sys.stderr,
            )
            await self._challenge(
                send, 403, "insufficient_scope", "The token lacks the required scope", self.base_scope
            )
            return

        print(
            f"[mcp-auth] 許可: sub={context.subject} scopes={list(context.scopes)}"
            f" token={context.fingerprint}",
            file=sys.stderr,
        )
        # ASGI では scope["state"] が下流への受け渡しに使われる慣習の場所
        state = scope.setdefault("state", {})
        state["auth"] = context
        await self.app(scope, receive, send)

    @staticmethod
    def _bearer_token(scope: Scope) -> str | None:
        for name, value in scope.get("headers", []):
            if name.lower() == b"authorization":
                text = value.decode("latin-1").strip()
                if text.lower().startswith("bearer "):
                    return text[7:].strip()
                return None
        return None

    async def _challenge(
        self,
        send: Send,
        status: int,
        error: str | None,
        description: str,
        scope_hint: str | None = None,
    ) -> None:
        # ⚠️ ヘッダーには ASCII しか置けない。日本語は本文の JSON に書く
        parts = [f'Bearer resource_metadata="{self.metadata_url}"']
        if error is not None:
            parts.append(f'error="{error}"')
        parts.append(f'error_description="{description}"')
        if scope_hint is not None:
            parts.append(f'scope="{scope_hint}"')
        message = (
            "このトークンには必要な権限（スコープ）がありません。"
            if status == 403
            else "有効なアクセストークンが必要です。WWW-Authenticate ヘッダーの resource_metadata を参照してください。"
        )
        body = json.dumps(
            {"jsonrpc": "2.0", "error": {"code": -32005 if status == 403 else -32004, "message": message}, "id": None},
            ensure_ascii=False,
        ).encode("utf-8")
        await self._send(send, status, body, [(b"www-authenticate", ", ".join(parts).encode("ascii"))])

    @staticmethod
    async def _send(
        send: Send, status: int, body: bytes, extra: list[tuple[bytes, bytes]]
    ) -> None:
        headers = [
            (b"content-type", b"application/json"),
            (b"content-length", str(len(body)).encode("ascii")),
            *extra,
        ]
        await send({"type": "http.response.start", "status": status, "headers": headers})
        await send({"type": "http.response.body", "body": body})

File: mcp_server_development/node/src/session12/auth-middleware.ts
/**
 * OAuth 2.1 の受付 ―― Protected Resource Metadata の公開とトークン検証
 *
 * セッション7 の `createMcpHttpEndpoint` が返す listener を包みます。
 * セッション7 の dispatch() に置いたコメント
 *   「★ セッション12（OAuth 2.1）では、この位置にアクセストークンの検証を 1 段足す」
 * と同じ順序（パス判定 → Origin/Host → トークン → セッション）になるように、
 * この層でも Origin/Host を先に見ています（8 節で理由を説明します）。
 */
import http from "node:http";

import {
  type TrustPolicy,
  createTrustPolicy,
  firstHeader,
  verifyRequestTrust,
} from "../session07/origin.js";
import { SCOPE_DOCS_READ, SUPPORTED_SCOPES, runWithAuth } from "./scopes.js";
import type { TokenVerifier, VerifyFailure } from "./token-verifier.js";

/**
 * RFC 9728。リソース識別子にパスがある場合は、well-known の「後ろ」にそのパスを足す。
 *   resource = http://127.0.0.1:3939/mcp
 *   → /.well-known/oauth-protected-resource/mcp
 * 互換のためパス無しの位置でも同じ内容を返します（仕様改訂で変わりうる箇所です）。
 */
export function protectedResourceMetadataPaths(endpoint: string): readonly string[] {
  const base = "/.well-known/oauth-protected-resource";
  return endpoint === "" || endpoint === "/" ? [base] : [base, `${base}${endpoint}`];
}

export type ProtectOptions = {
  /** セッション7 の createMcpHttpEndpoint().listener */
  readonly inner: http.RequestListener;
  readonly verifier: TokenVerifier;
  /** このサーバー自身の識別子。トークンの aud はこれと完全一致しなければならない */
  readonly resource: string;
  readonly authorizationServers: readonly string[];
  readonly endpoint?: string;
  readonly trust?: TrustPolicy;
  /** このリソースに触るための最低スコープ */
  readonly baseScope?: string;
};

type Challenge = {
  readonly status: number;
  readonly error?: string;
  /** ⚠️ ヘッダーには ASCII しか置けない。日本語を入れると ERR_INVALID_CHAR で落ちる */
  readonly description: string;
  readonly scope?: string;
};

/** 検証失敗の理由 → 返すべき HTTP の応答。理由コードはログ用、文面は無情報 */
const CHALLENGES: Readonly<Record<VerifyFailure, Challenge>> = {
  missing_token: { status: 401, description: "An access token is required" },
  malformed_header: {
    status: 401,
    error: "invalid_request",
    description: "The Authorization header is malformed",
  },
  malformed_token: { status: 401, error: "invalid_token", description: "The token is malformed" },
  unsupported_alg: {
    status: 401,
    error: "invalid_token",
    description: "The signature algorithm is not allowed",
  },
  embedded_key: {
    status: 401,
    error: "invalid_token",
    description: "The token header carries key material",
  },
  unknown_kid: { status: 401, error: "invalid_token", description: "The signing key is unknown" },
  bad_signature: { status: 401, error: "invalid_token", description: "The signature is invalid" },
  issuer_mismatch: { status: 401, error: "invalid_token", description: "The issuer is not trusted" },
  expired: { status: 401, error: "invalid_token", description: "The access token is expired" },
  not_yet_valid: { status: 401, error: "invalid_token", description: "The token is not valid yet" },
  audience_mismatch: {
    status: 401,
    error: "invalid_token",
    description: "The token was not issued for this resource",
  },
  jwks_unavailable: {
    status: 503,
    error: "temporarily_unavailable",
    description: "Signing keys are temporarily unavailable",
  },
};

export function protectWithOAuth(options: ProtectOptions): http.RequestListener {
  const endpoint = options.endpoint ?? "/mcp";
  const trust = options.trust ?? createTrustPolicy();
  const baseScope = options.baseScope ?? SCOPE_DOCS_READ;
  const metadataPaths = protectedResourceMetadataPaths(endpoint);
  const canonicalPath = metadataPaths[metadataPaths.length - 1] ?? metadataPaths[0] ?? "/";
  const metadataUrl = `${new URL(options.resource).origin}${canonicalPath}`;
  const metadataBody = JSON.stringify({
    resource: options.resource,
    authorization_servers: [...options.authorizationServers],
    scopes_supported: [...SUPPORTED_SCOPES],
    bearer_methods_supported: ["header"],
    resource_name: "社内ドキュメント検索 MCP サーバー",
  });

  async function handle(req: http.IncomingMessage, res: http.ServerResponse): Promise<void> {
    const pathname = new URL(req.url ?? "/", "http://placeholder").pathname;

    // ① メタデータは認証の前に置く。401 を受けた人がここを読みに来るのだから、
    //    ここを保護したら発見フローが永久に閉じる（鶏と卵になる）
    if (metadataPaths.includes(pathname)) {
      if (req.method !== "GET") {
        res.writeHead(405, { allow: "GET", "content-type": "application/json" });
        res.end(JSON.stringify({ error: "method_not_allowed" }));
        return;
      }
      res.writeHead(200, {
        "content-type": "application/json",
        "content-length": Buffer.byteLength(metadataBody),
        // 変わらない情報なのでキャッシュさせる。毎リクエスト取りに来られると無駄
        "cache-control": "public, max-age=3600",
      });
      res.end(metadataBody);
      return;
    }

    // ② Origin / Host（セッション7 の受付）。下流でも走るが多層防御として残す
    const verdict = verifyRequestTrust(req.headers, trust);
    if (!verdict.ok) {
      authLog(`拒否（${verdict.reason}）: origin=${firstHeader(req.headers.origin) ?? "(なし)"}`);
      sendJson(res, 403, -32003, verdict.detail);
      return;
    }

    // ③ アクセストークンの検証（セッション7 が「ここに 1 段足す」と書いた位置）
    const outcome = await options.verifier.verify(firstHeader(req.headers.authorization));
    if (!outcome.ok) {
      // 未知のパスでも 404 を返さない。認証を通す前に「存在するパス」を教えないため
      authLog(`認証失敗（${outcome.reason}）: ${req.method ?? "?"} ${pathname}`);
      sendChallenge(res, CHALLENGES[outcome.reason], metadataUrl);
      return;
    }
    const context = outcome.context;

    // ④ ベーススコープ。ここは「接続してよいか」の判定なので HTTP 層で 403 を返す
    if (!context.scopes.includes(baseScope)) {
      authLog(
        `認可失敗（insufficient_scope）: sub=${context.subject} scopes=[${context.scopes.join(" ")}]` +
          ` token=${context.fingerprint}`,
      );
      sendChallenge(
        res,
        {
          status: 403,
          error: "insufficient_scope",
          description: "The token lacks the required scope",
          scope: baseScope,
        },
        metadataUrl,
      );
      return;
    }

    authLog(
      `許可: sub=${context.subject} client=${context.clientId ?? "-"}` +
        ` scopes=[${context.scopes.join(" ")}] token=${context.fingerprint}`,
    );
    // ⑤ 認証文脈を非同期スコープに載せて下流（MCP 層）へ渡す
    runWithAuth(context, () => options.inner(req, res));
  }

  return (req, res) => {
    handle(req, res).catch((error: unknown) => {
      authLog(`認証層で未処理の例外: ${error instanceof Error ? error.message : String(error)}`);
      if (res.headersSent) {
        res.end();
        return;
      }
      sendJson(res, 500, -32603, "サーバー内部エラーです。");
    });
  };
}

function sendChallenge(res: http.ServerResponse, challenge: Challenge, metadataUrl: string): void {
  // WWW-Authenticate の組み立て。auth-scheme のあとにパラメータをカンマで並べる
  const parts = [`Bearer resource_metadata="${metadataUrl}"`];
  if (challenge.error !== undefined) {
    parts.push(`error="${challenge.error}"`);
  }
  parts.push(`error_description="${challenge.description}"`);
  if (challenge.scope !== undefined) {
    parts.push(`scope="${challenge.scope}"`);
  }
  const message =
    challenge.status === 403
      ? "このトークンには必要な権限（スコープ）がありません。必要なスコープは WWW-Authenticate ヘッダーを参照してください。"
      : challenge.status === 503
        ? "署名鍵を取得できないため、いまトークンを検証できません。しばらく待って再試行してください。"
        : "有効なアクセストークンが必要です。WWW-Authenticate ヘッダーの resource_metadata から認可サーバーを見つけ、トークンを取得してください。";
  const body = JSON.stringify({
    jsonrpc: "2.0",
    error: { code: challenge.status === 403 ? -32005 : -32004, message },
    id: null,
  });
  const headers: Record<string, string> = {
    "www-authenticate": parts.join(", "),
    "content-type": "application/json",
    "content-length": String(Buffer.byteLength(body)),
  };
  if (challenge.status === 503) {
    headers["retry-after"] = "5";
  }
  res.writeHead(challenge.status, headers);
  res.end(body);
}

function sendJson(res: http.ServerResponse, status: number, code: number, message: string): void {
  const body = JSON.stringify({ jsonrpc: "2.0", error: { code, message }, id: null });
  res.writeHead(status, {
    "content-type": "application/json",
    "content-length": Buffer.byteLength(body),
  });
  res.end(body);
}

export function authLog(message: string): void {
  console.error(`[mcp-auth] ${message}`);
}
